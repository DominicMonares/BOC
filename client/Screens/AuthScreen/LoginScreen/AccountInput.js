import { useState } from 'react';
import { Pressable, Text, TextInput, View } from 'react-native';
import { connect, useSelector, useDispatch } from 'react-redux';
import axios from 'axios';
// import { NavigationContainer } from '@react-navigation/native';
// import { createNativeStackNavigator } from '@react-navigation/native-stack';
//import { useNavigation } from '@react-navigation/native';


import styles from '../Styles';
import { API_IP_login } from '../../../constants.js';
import { login } from '../../../Redux/actions';
import { authLog } from '../../../Redux/actions';
import { authReg } from '../../../Redux/actions';
import { palette } from '../../../Utils/ColorScheme';


//import { authlog } from '../../../Redux/actions';

const loginEndpoint = `http://${API_IP_login}/user/login/password`;

export default function AccountInput({ navigation }){
  const state = useSelector(state => state);
  const dispatch = useDispatch();
  const screen = useSelector(state => state.authScreen);
  let theme = palette(state.theme);


  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');

  const handleLogin = async () => {
     console.log('endpoint', loginEndpoint);
     console.log('username', username);
     console.log('password', password);


  await axios.post(loginEndpoint, {
    username: username,
    password: password
  })
  .then(function (response) {

    console.log('successfully sent login data to backend');
    dispatch(login(username));

  })
  .catch(function (error) {
    console.log('error sending login', error);
    console.log(error.response.data)
  });


}




  const handleSignUpRedirect= async () => {
    console.log('handleSignUpRedirect was called');
    dispatch(authReg());


  }

  return (
    <View style={styles.loginFields}>
      <Text style={styles.fieldLabels}>Account name</Text>
      <TextInput
        style={[{ borderColor: theme.buttonBorderColor }, styles.field]}
        onChangeText={text => setUsername(text)}
        autoCapitalize="none"
      />
      <Text style={styles.fieldLabels}>Password</Text>
      <TextInput
        style={[{ borderColor: theme.buttonBorderColor }, styles.field]}
        secureTextEntry={true}
        onChangeText={text => setPassword(text)}

      />

      <View style={styles.buttonContainer}>
        <Pressable
          title={'Continue'}
          style={[
            { backgroundColor: theme.buttonColor, borderColor: theme.buttonBorderColor },
            styles.button
          ]}
          onPress={() => handleLogin()}>
          <Text
            style={[{ color: theme.tabIconInactive }, styles.buttonText]}>
            Log In
          </Text>
        </Pressable>
      </View>

    </View>
  )




}

