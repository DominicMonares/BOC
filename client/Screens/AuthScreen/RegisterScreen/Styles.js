import { StyleSheet } from 'react-native';

const Register = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
  fields: {
    // backgroundColor: 'grey',
    width: '70%'
  },
  field: {
    backgroundColor: 'white',
    width: '100%',
    height: '10%'
  }
})

export { Register };