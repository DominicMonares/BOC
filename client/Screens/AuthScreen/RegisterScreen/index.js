import { Text, View, TouchableOpacity } from 'react-native';
import API_IP from '../../../constants.js';

import axios from 'axios';

export default function RegisterScreen() {
  const onRegisterPress = () => {
    console.log('TEST')
    axios.post(`http://${API_IP}/addNewUser`, {
      test: 1,
      tests: 2
    }, {
      headers: {
        'Content-Type': 'text/json',
        'Accept': 'application/json'
      }
    })
    .then(res => {console.log(res)})
    .catch(err => {console.log(err)});
  }

  return (
    <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
      <TouchableOpacity onPress={() => onRegisterPress()}>
        <Text>Register Screen</Text>
      </TouchableOpacity>
    </View>
  );
}