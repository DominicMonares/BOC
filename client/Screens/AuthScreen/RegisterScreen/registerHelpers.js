const containsUpperCase = (pw) => {
  let chars = pw.split('');
  return chars.some(char => {
    if (char < char.toLowerCase()) {
      return true;
    } else {
      return false;
    }
  })
}

const containsNumber = (pw) => {
  let chars = pw.split('');
  return chars.some(char => {
    if (Number(char) || Number(char) === 0) {
      return true;
    } else {
      return false;
    }
  })
}

const containsSpecial = (pw) => {
  let special = false;
  const specialChars = /[`!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?~]/;
  return specialChars.test(pw);
}

module.exports = { containsUpperCase, containsNumber, containsSpecial };
