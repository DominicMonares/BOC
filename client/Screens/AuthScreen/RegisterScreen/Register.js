// React | Redux
import { useState } from 'react';
import { Button, Text, TextInput, View } from 'react-native';
import { useSelector } from 'react-redux';

// Components
import Header from '../Shared/Header';
import LogoBackButton from '../Shared/LogoBackButton';
import Tabs from '../Shared/Tabs';
import BackButton from '../Shared/BackButton';
import Google from '../Shared/Google';
import Twitter from '../Shared/Twitter';
import AccountInput from './components/AccountInput';
import NavBar from '../Shared/NavBar';

// Styling
import { Register as styles } from './Styles'
import { lightTheme, darkTheme } from '../../../constants';


const RegisterScreen = function() {
  const theme = useSelector(state => state.theme);

  var current;
  if (theme) {
    current = lightTheme;
  } else {
    current = darkTheme;
  }

  return (
    <View style={[{ backgroundColor: current.pageColor }, styles.container]}>
      <Header />
      <LogoBackButton />
      <Tabs />
      <AccountInput />
      <Google />
      <Twitter />
      <View style={styles.bottom}>
        <NavBar />
      </View>
    </View>
  )
}

export default RegisterScreen;
