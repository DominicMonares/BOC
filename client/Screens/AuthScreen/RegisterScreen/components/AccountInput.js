import { useState } from 'react';
import { Pressable, Text, TextInput, View } from 'react-native';
import { useSelector, useDispatch } from 'react-redux';
import axios from 'axios';

import styles from '../../Styles';
import { API_IP } from '../../../../constants.js';
import { containsUpperCase, containsNumber, containsSpecial } from '../registerHelpers';
import { authLog } from '../../../../Redux/actions';
import { lightTheme, darkTheme } from '../../../../constants';
import { useFonts } from "expo-font";

const registrationEndpoint = `http://${API_IP}/user/addNewUser`;

export default function AccountInput() {
  const state = useSelector(state => state);
  const screen = useSelector(state => state.authScreen);
  const theme = useSelector(state => state.theme);
  const dispatch = useDispatch();

  const [username, setUsername] = useState('');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [confirmPw, setConfirmPw] = useState('');
  const [invalidUsername, setInvalidUsername] = useState(<></>);
  const [invalidEmail, setInvalidEmail] = useState(<></>);
  const [invalidPassword, setInvalidPassword] = useState(<></>);
  const [passwordLength, setPasswordLength] = useState(<></>);
  const [passwordCaptial, setPasswordCapital] = useState(<></>);
  const [passwordNum, setPasswordNum] = useState(<></>);
  const [passwordSpecial, setPasswordSpecial] = useState(<></>);
  const [current, setCurrent] = useState(() => {
    if (theme) {
      return lightTheme;
    } else {
      return darkTheme;
    }
  });

  const [fontsLoaded] = useFonts({
    comicSans: require('../../../../assets/fonts/comic.ttf')
  });

  const handleSignUp = async () => {

    let validatedUsername = validUsername(username);
    let validatedEmail = validEmail(email);
    let validatedPasswords = validPassword(password, confirmPw);
    let validatedPwStrength = strongPassword(password);

    if (validatedUsername && validatedEmail && validatedPasswords && validatedPwStrength) {
      try {
        // Must use HTTPS when deployed in order to keep pw safe
        const register = await axios.post(registrationEndpoint, {
          username: username.toLowerCase(),
          email: email,
          password: password
        });

        dispatch(authLog());
        alert('We\'ve sent you an activation email!\nPlease check your inbox.');
      } catch (err) {
        console.log(err);
      }
    }
  }

  const validUsername = (username) => {
    let letters = username.split('');

    if (username.length >= 2 && username.length <= 16) {
      setInvalidUsername(<></>);
      return true;
    } else {
      setInvalidUsername(<Text>Username must be between 2 and 16 characters long.</Text>);
      return false;
    }
  }

  const validEmail = (email) => {
    if (email.split('@').length === 2) {
      let splitEmail = email.split('@');
      let left = splitEmail[0].split('.').length;
      let right = splitEmail[1].split('.').length;
      let leftMatch = left <= 2 && left > 0;
      let rightMatch = right <= 2 && right > 0;

      if (leftMatch && rightMatch) {
        setInvalidEmail(<></>);
        return true;
      } else {
        setInvalidEmail(<Text>Invalid email format.</Text>);
        return false;
      }
    } else {
      setInvalidEmail(<Text>Invalid email format.</Text>);
      return false;
    }
  }

  const validPassword = (pw1, pw2) => {
    if (pw1 === pw2) {
      setInvalidPassword(<></>)
      return true;
    } else {
      setInvalidPassword(<Text>Passwords do not match.</Text>)
      return false;
    }
  }

  const strongPassword = (pw) => {
    let length = pw.length >= 8;
    let capital = containsUpperCase(pw);
    let num = containsNumber(pw);
    let special = containsSpecial(pw);

    if (!length) {
      setPasswordLength(<Text>Password must contain at least 8 characters.</Text>);
    } else {
      setPasswordLength(<></>);
    }

    if (!capital) {
      setPasswordCapital(<Text>Password must contain at least 1 capital letter.</Text>);
    } else {
      setPasswordCapital(<></>);
    }

    if (!num) {
      setPasswordNum(<Text>Password must contain at least 1 number.</Text>);
    } else {
      setPasswordNum(<></>);
    }

    if (!special) {
      setPasswordSpecial(<Text>Password must contain at least 1 special character.</Text>);
    } else {
      setPasswordSpecial(<></>);
    }

    if (length && capital && num && special) {
      return true;
    }
  }

  return (
    <View style={styles.fields}>
      <Text style={styles.fieldLabels}>Username</Text>
      <TextInput
        style={[{ borderColor: current.buttonBorderColor }, styles.field]}
        accessibilityLabel="reg-username"
        onChangeText={text => setUsername(text)}
        autoCapitalize="none"
      />
      <Text style={styles.fieldLabels}>Email Address</Text>
      <TextInput
        style={[{ borderColor: current.buttonBorderColor }, styles.field]}
        accessibilityLabel="reg-email"
        onChangeText={text => setEmail(text)}
        autoCapitalize="none"
      />
      <Text style={styles.fieldLabels}>Password</Text>
      <TextInput
        style={[{ borderColor: current.buttonBorderColor }, styles.field]}
        accessibilityLabel="reg-pw1"
        onChangeText={text => setPassword(text)}
        secureTextEntry={true}
        textContentType="oneTimeCode"
      />
      <Text style={styles.fieldLabels}>Password Again</Text>
      <TextInput
        style={[{ borderColor: current.buttonBorderColor }, styles.field]}
        accessibilityLabel="reg-pw2"
        onChangeText={text => setConfirmPw(text)}
        secureTextEntry={true}
        textContentType="oneTimeCode"
      />
      <View style={styles.buttonContainer}>
        <Pressable
          title={'Sign Up'}
          style={[
            { backgroundColor: current.buttonColor, borderColor: current.buttonBorderColor },
            styles.button
          ]}
          onPress={() => handleSignUp()}>
          <Text
            style={[{ color: current.tabIconInactive }, styles.buttonText]}>
            Sign Up
          </Text>
        </Pressable>
      </View>
      {invalidUsername}
      {invalidEmail}
      {passwordLength}
      {passwordCaptial}
      {passwordNum}
      {passwordSpecial}
      {invalidPassword}
    </View>
  )
}
