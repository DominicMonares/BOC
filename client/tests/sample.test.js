/**
 * @jest-environment jsdom
 */

import React from 'react';
import { Provider } from 'react-redux';
import { create } from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react-native';
import configureStore from 'redux-mock-store';

import AppWrapper from '../App';
import RegisterScreen from '../Screens/AuthScreen/RegisterScreen/Register.js';
import AccountInput from '../Screens/AuthScreen/RegisterScreen/components/AccountInput';

const mockStore = configureStore();

describe('Samples', () => {

  describe('@testing-library/react-native', () => {
    const state = {
      authScreen: 'register',
      guestHome: 'home',
      theme: true,
      user: {
        username: 'bigglesworth'
      }
    };

    let store, acctInput;

    beforeEach(() => {
      store = mockStore(state);
      acctInput = <Provider store={store}>
        <AccountInput />
      </Provider>
    })

    it('should render and validate username field', () => {
      const { getByLabelText, getByText, toJSON } = render(acctInput);
      const children = toJSON().children;
      const username = children[0]['children'][0];
      expect(username).toBe('Username');

      fireEvent.changeText(getByLabelText('reg-username'), 'bigglesworth');
      fireEvent.press(getByText('Sign Up'));
      expect(toJSON()['children'][9]['children'][0]).toBe('Invalid email format.');
    })

  })

  describe('react-test-renderer', () => {

    // Unit Test - Replaced by React Testing Library .toJSON
    it('should render AppWrapper', () => {
      const tree = renderer.create(<AppWrapper />).toJSON();

      // AppWrapper renders UnAuthorized page by default
      expect(tree.children.length).toBe(2);
    })

    // Unit Test - Replaced by React Testing Library .toJSON
    it('should render guest discovery feed', () => {
      const tree = renderer.create(<AppWrapper />).toJSON();
      const feed = tree['children'][0]['children'][0];

      // placeholder text from UnAuthorized page
      expect(feed).toBe('Discovery Feed Here');
    })

    // Snapshot Test
    it('should render app wrapper', () => {
      const tree = create(<AppWrapper />).toJSON();
      expect(tree).toMatchSnapshot();
    })

  })

})