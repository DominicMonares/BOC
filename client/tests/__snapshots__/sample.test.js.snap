// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render app wrapper 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Discovery Feed Here
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Login Icon Here
      </Text>
    </View>
  </View>
</View>
`;
